### PR TITLE
Deltas buffer

### PR DESCRIPTION
**Pull request by kayibal**

### CODE REVIEW

File: tycho-core/src/models/contract.rs [hunk 1]
```diff
@@ -52,6 +52,29 @@ impl Contract {
...
         self.balance_modify_tx = modified_at.clone();
     }
+
+    pub fn apply_contract_delta(&mut self, delta: &ContractDelta) -> Result<(), DeltaError> {
+        let self_id = (self.chain, &self.address);
+        let other_id = (delta.chain, &delta.address);
+        if self_id != other_id {
+            return Err(DeltaError::IdMismatch(format!("{:?}", self_id), format!("{:?}", other_id)));
+        }
+        if let Some(balance) = delta.balance.as_ref() {
```
Comment (zizou0x) [inline tycho-core/src/models/contract.rs:62]: TODO for `balance_modify_tx`, `code_modify_tx` and `code_hash`

File: tycho-indexer/src/extractor/revert_buffer.rs [hunk 2]
```diff
@@ -1,12 +1,51 @@
...
 };
 
+#[derive(Clone, Debug, Copy)]
+pub enum BlockNumberOrTimestamp {
+    Number(u64),
+    Timestamp(NaiveDateTime),
+}
+
+impl BlockNumberOrTimestamp {
+    fn greater_then(&self, other: &Block) -> bool {
```
Comment (zizou0x) [inline tycho-indexer/src/extractor/revert_buffer.rs]:
```suggestion
    fn greater_than(&self, other: &Block) -> bool {
```

Comment (louise-poole) [inline tycho-indexer/src/extractor/revert_buffer.rs]:
```suggestion
    fn greater_than(&self, other: &Block) -> bool {
```

File: tycho-indexer/src/services/deltas_buffer.rs [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+            (Some(msg), None) => {
+                if let Some(buffer) = self.native.get(&msg.extractor) {
+                    let mut guard = buffer.lock().map_err(|e| {
+                        PendingDeltasError::LockError("Native".to_string(), e.to_string())
+                    })?;
+                    if msg.revert {
+                        guard.purge(msg.block.hash.clone())?;
+                    } else {
+                        guard.insert_block(msg.clone())?;
+                        // TODO: remove corresponding final block
```
Comment (louise-poole) [inline tycho-indexer/src/services/deltas_buffer.rs]: ahhhhhhhh

File: tycho-indexer/src/services/deltas_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+                .lock()
+                .map_err(|e| PendingDeltasError::LockError("VM".to_string(), e.to_string()))?;
+            for entry in guard.get_block_range(None, version)? {
+                if let Some(delta) = entry.deltas.get(&db_state.address) {
+                    db_state.apply_contract_delta(delta)?;
+                    change_found = true;
+                }
+
+                // TODO: currently it is impossible to apply balance changes since
+                //  we don't know the component_id of the contract.
```
Comment (louise-poole) [inline tycho-indexer/src/services/deltas_buffer.rs:192]: Boo. Design flaw? We should ideally track balances per contract no?

File: tycho-core/src/models/protocol.rs [hunk 5]
```diff
@@ -62,6 +66,38 @@ impl ProtocolComponentState {
...
+            ));
+        }
+        self.attributes
+            .extend(delta.updated_attributes.clone());
+
+        self.attributes
+            .retain(|attr, _| !delta.deleted_attributes.contains(attr));
+
+        Ok(())
+    }
```
Comment (louise-poole) [inline tycho-core/src/models/protocol.rs:90]: We should document that we assume that the passed delta is newer than self and therefore will overwrite common attributes.

File: tycho-core/src/models/protocol.rs [hunk 6]
```diff
@@ -62,6 +66,38 @@ impl ProtocolComponentState {
...
+        delta: &HashMap<Bytes, ComponentBalance>,
+    ) -> Result<(), DeltaError> {
+        self.balances.extend(
+            delta
+                .iter()
+                .map(|(k, v)| (k.clone(), v.new_balance.clone())),
+        );
+
+        Ok(())
+    }
```
Comment (louise-poole) [inline tycho-core/src/models/protocol.rs:106]: Same here, we assume that the passed balance is a newer balance.

File: tycho-indexer/src/services/deltas_buffer.rs [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+    }
+
+    #[test]
+    fn test_insert_native() {
+        let buffer = PendingDeltas::new([], ["native:extractor"]);
+        let exp: NativeBlockDeltas = (&native_block_deltas()).into();
+
+        buffer
+            .insert(Arc::new(native_block_deltas()))
+            .expect("vm insert failed");
```
Comment (louise-poole) [inline tycho-indexer/src/services/deltas_buffer.rs]: Copy pasta

File: tycho-indexer/src/services/deltas_buffer.rs [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+
+        let native_revert_buffer = buffer
+            .native
+            .get("native:extractor")
+            .expect("vm:extractor buffer missing");
```
Comment (louise-poole) [inline tycho-indexer/src/services/deltas_buffer.rs]: Copy pasta

File: tycho-indexer/src/services/mod.rs [hunk 9]
```diff
@@ -115,8 +132,26 @@ where
         struct ApiDoc;
 
         let openapi = ApiDoc::openapi();
+        let pending_deltas = PendingDeltas::new(
+            self.native_extractors
+                .iter()
+                .map(String::as_str),
+            self.native_extractors
+                .iter()
+                .map(String::as_str),
```
Comment (louise-poole) [inline tycho-indexer/src/services/mod.rs:141]: vm_extractors?

**Pull request merged**